File: lab5/haskell/prime_factors.hs
import Text.Printf


factorize :: Int -> Int -> [Int]
factorize 1 _ = []
factorize n divisor
    | n `mod` divisor == 0 = divisor : factorize (n `div` divisor) divisor
    | otherwise            = factorize n (divisor + 1)

primeFactors :: Int -> [Int]
primeFactors n
    | n <= 1    = []
    | otherwise = factorize n 2


printPrimeFactors :: Int -> IO ()
printPrimeFactors n = do
    printf "Prime factors of %d: %s\n" n (show $ primeFactors n)


main :: IO ()
main = do
    printPrimeFactors 17
    printPrimeFactors 36

File: lab5/prolog/prime_factors.pl
factorize(1, _, []).
factorize(N, Divisor, [Divisor|Factors]) :-
    N mod Divisor =:= 0,
    NewN is N // Divisor,
    factorize(NewN, Divisor, Factors).
factorize(N, Divisor, Factors) :-
    N mod Divisor =\= 0,
    NextDivisor is Divisor + 1,
    factorize(N, NextDivisor, Factors).

primeFactors(N, Factors) :-
    N =< 1,
    Factors = [].
primeFactors(N, Factors) :-
    N > 1,
    factorize(N, 2, Factors).


printPrimeFactors(N) :-
    primeFactors(N, Factors),
    format("Prime factors of ~d: ~w~n", [N, Factors]).


main :-
    printPrimeFactors(17),
    printPrimeFactors(36).

:- main.
